test/tests/functional/pbs_qrun.py

```diff
@@ -92,3 +92,54 @@ def test_invalid_host_val(self):
         msg = "Server is not up"
         self.assertTrue(self.server.isUp(), msg)
         self.logger.info("As expected server is up and running")
+
+    def test_qrun_hangs(self):
+        """
+        This test submit 500 jobs with differnt equivalence class,
+        turn of scheduling and qrun job to
+        verify whether qrun hangs.
+        """
+        node = self.mom.shortname
+        err_msg = 'qrun: Not enough free nodes available'
+        joblist = []
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'scheduling': 'False'})
+        self.server.manager(MGR_CMD_SET, NODE,
+                            {'resources_available.ncpus': 1}, id=node)
+        for walltime in range(1, 501):
+            j = Job(TEST_USER)
+            j.create_script('/bin/date')
+            a = {'Resource_List.walltime': walltime}
+            j.set_attributes(a)
+            jid = self.server.submit(j)
+            joblist.append(jid)
+        self.logger.info("Submitted 500 jobs with different walltime")
+
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'scheduling': 'True'})
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'scheduling': 'False'})
+        job = 10  # when we turn on scheduling, it scheduled few job.
+        while job < 501:
+            jid = joblist[job]
+            now = int(time.time())
+            try:
+                self.server.runjob(jobid=jid)
+            except PbsRunError as e:
+                # sched cycle started by qrun and failed as no free nodes
+                if err_msg in e.msg[0]:
+                    self.logger.info(
+                        "No free nodes available. Checking for sched cycle")
+                    try:
+                        self.scheduler.log_match("Starting Scheduling Cycle",
+                                                 starttime=now,
+                                                 max_attempts=10)
+                    except PtlLogMatchError:
+                        self.assertFalse(True,
+                                         "Sched cycle not started by qrun")
+                    else:
+                        continue
+            except Exception as e:
```

Comment:
I just looked at self.server.runjob() and it won't raise an exception other than PbsRunError.  If this triggers, won't the self.server.runjob() hang?  Will this try/except do what you want it to do?

Reply:
Yes.
When this "except Exception as e" triggers, that mean qrun hang. If we don't handle this in except block, then it will raise Timeout exception as below (unhandled exception caused stack unwinding)

2018-12-06 09:36:07,546 INFO     TIMEDOUT
2018-12-06 09:36:07,546 INFO     ==========================
2018-12-06 09:36:07,546 INFO      Entered TestQrun tearDown
2018-12-06 09:36:07,546 INFO     ==========================
2018-12-06 09:36:07,547 INFO     ===========================
2018-12-06 09:36:07,547 INFO     Completed TestQrun tearDown
2018-12-06 09:36:07,547 INFO     ===========================
2018-12-06 09:36:07,549 INFO     
2018-12-06 09:36:07,549 INFO     ======================================================================
2018-12-06 09:36:07,549 INFO     TIMEDOUT: test_qrun_hangs (tests.functional.pbs_qrun.TestQrun)

2018-12-06 09:36:07,550 INFO     ___m_oo_m___
2018-12-06 09:36:07,550 INFO     Traceback (most recent call last):
  File "/home/sudemo/work/golden/pbspro/test/tests/functional/pbs_qrun.py", line 124, in test_qrun_hangs
    self.server.runjob(jobid=jid)
  File "/usr/lib/python2.7/site-packages/ptl/lib/pbs_testlib.py", line 7269, in runjob
    logerr=logerr)
  File "/usr/lib/python2.7/site-packages/ptl/utils/pbs_dshutils.py", line 951, in run_cmd
    (o, e) = p.communicate(input)
  File "/usr/lib64/python2.7/subprocess.py", line 800, in communicate
    return self._communicate(input)
  File "/usr/lib64/python2.7/subprocess.py", line 1401, in _communicate
    stdout, stderr = self._communicate_with_poll(input)
  File "/usr/lib64/python2.7/subprocess.py", line 1455, in _communicate_with_poll
    ready = poller.poll()
  File "/usr/lib/python2.7/site-packages/ptl/utils/plugins/ptl_test_runner.py", line 602, in timeout_handler
    raise TimeOut('Timed out after %s second' % timeout)
TimeOut: Timed out after 180 second

Reply:
Oh, I see what is going on.  The test case times out from the PTL timeout.  You're catching that exception to cause the test to fail.  Nice.